docs/udf-guide.md

```diff
@@ -0,0 +1,159 @@
+    str => $"{s1} {str}");
+
+```
+
+And for a sample Dataframe, let's take the following Dataframe `df`:
+
+```text
+    +-------+
+    |   name|
+    +-------+
+    |Michael|
+    |   Andy|
+    | Justin|
+    +-------+
+```
+
+Now let's apply the above defined `udf` to the dataframe `df`:
+
+```csharp
+DataFrame udfResult = df.Select(udf(df["name"]));
+```
+
+This would return the below as the Dataframe `udfResult`:
+
+```text
+    +-------------+
+    |         name|
+    +-------------+
+    |hello Michael|
+    |   hello Andy|
+    | hello Justin|
+    +-------------+
+```
+
+## Good to know while implementing UDFs
+
+One behavior to be aware of while implementing UDFs in .NET for Apache Spark is the way target of the UDF gets serialized. .NET for Apache Spark uses .NET Core which does not support serializing delegates, so it is instead done by using reflection to serialize the target where the delegate is defined. When multiple delegates are defined in a common scope, they have a shared closure that becomes the target of reflection for serialization. Let's take an example to illustrate what that means:
+
+Sample user code:
+
+```csharp
+using System;
+
+public class C {
+    public void M() {
+        string s1 = "s1";
+        string s2 = "s2";
+        Func<string, string> a = str => s1;
+        Func<string, string> b = str => s2;
+    }
+}
+```
+
+The above C# code generates the following IL (Intermediate language) code from the compiler:
+
+```csharp
+public class C
+{
+    [CompilerGenerated]
+    private sealed class <>c__DisplayClass0_0
+    {
+        public string s1;
+
+        public string s2;
+
+        internal string <M>b__0(string str)
+        {
+            return s1;
+        }
+
+        internal string <M>b__1(string str)
+        {
+            return s2;
+        }
+    }
+
+    public void M()
+    {
+        <>c__DisplayClass0_0 <>c__DisplayClass0_ = new <>c__DisplayClass0_0();
+        <>c__DisplayClass0_.s1 = "s1";
+        <>c__DisplayClass0_.s2 = "s2";
+        Func<string, string> func = new Func<string, string>(<>c__DisplayClass0_.<M>b__0);
+        Func<string, string> func2 = new Func<string, string>(<>c__DisplayClass0_.<M>b__1);
+    }
+}
+```
+As can be seen in the above IL code, both `func` and `func2` share the same closure `<>c__DisplayClass0_0`, which is the target that is serialized when serializing the delegates `func` and `func2`. Hence, even though `Func<string, string> a` is only referencing `s1`, `s2` also gets serialized when sending over the bytes to the workers.
+This can lead to some unexpected behaviors at runtime (for example in the case of using Broadcast variables, as explained in more detail in [this guide](broadcast-guide.md)), which is why we recommend restricting the visibility of the variables used in a function to that function's scope.
+Taking the above example to better explain what that means:
+
+Recommended user code:
+
+```csharp
```

Comment:
Is this the same recommended user code as above? If so, rather than repeating it (which will make the doc longer and could be a bit confusing), I think it'd make sense just to state that you're using the same user code to accomplish task X, as was shown above.

Reply:
No actually this is not supposed to be the same code (my bad), thanks for pointing that out! Will make the change.